File: priors-research/priors/posterior.py
"""Grid posterior over (θ_e, θ_i, β). SPEC §3.4.

    P(engage | p, θ, β) = 1 / (1 + exp(β · (p − θ)))

Joint over 33 × 25 × 15 = 12,375 cells. Multiplicative update after every
decision, renormalise. That is the whole model.

Everything is done in log space: β reaches 30 and prices reach 0.85, so
β·(p−θ) reaches ~24 and a naive exp() loses precision well before that.

This module is the reference for `PriorsEngine/Posterior.swift`, which must
reproduce it on a fixed decision sequence. Keep the two structurally aligned.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable, Sequence

import numpy as np

from priors.scenarios import BETA, THETA_E, THETA_I, GridSpec, TraitName


def build_grid(spec: GridSpec) -> np.ndarray:
    """Materialise a parameter axis. Log-spaced axes use geomspace (SPEC §3.3)."""
    if spec.log:
        return np.geomspace(spec.lo, spec.hi, spec.n, dtype=np.float64)
    return np.linspace(spec.lo, spec.hi, spec.n, dtype=np.float64)


@dataclass(frozen=True)
class PosteriorSnapshot:
    """SCHEMA §4. Plain values only — safe to serialise straight to JSON."""

    theta_e_mean: float
    theta_e_sd: float
    theta_e_grid: list[float]
    theta_e_marginal: list[float]
    theta_i_mean: float
    theta_i_sd: float
    theta_i_grid: list[float]
    theta_i_marginal: list[float]
    beta_mean: float
    beta_sd: float


def log_choice_lik(
    price: float,
    engaged: bool,
    theta: np.ndarray,
    beta: np.ndarray,
) -> np.ndarray:
    """log P(response | price, θ, β) for the SPEC §3.4 choice model.

    ``-logaddexp(0, z)`` is log(sigmoid(-z)), evaluated without overflow.
    """
    z = beta * (price - theta)
    return -np.logaddexp(0.0, z) if engaged else -np.logaddexp(0.0, -z)


class Posterior:
    """Joint posterior over (θ_e, θ_i, β).

    Axis order is fixed and load-bearing: ``(theta_e, theta_i, beta)``.
    A θ_e decision is constant along axis 1; a θ_i decision along axis 0.
    """

    __slots__ = ("theta_e", "theta_i", "beta", "_log_post", "_te", "_ti", "_b")

    def __init__(self) -> None:
        self.theta_e = build_grid(THETA_E)
        self.theta_i = build_grid(THETA_I)
        self.beta = build_grid(BETA)

        # Broadcast views, shaped for the joint. Built once.
        self._te = self.theta_e[:, None, None]
        self._ti = self.theta_i[None, :, None]
        self._b = self.beta[None, None, :]

        # SPEC §3.1/§3.2 — uniform prior over each grid's support.
        shape = (THETA_E.n, THETA_I.n, BETA.n)
        self._log_post = np.full(shape, -np.log(float(np.prod(shape))), dtype=np.float64)

    # -- core ---------------------------------------------------------------

    @property
    def log_post(self) -> np.ndarray:
        return self._log_post

    @property
    def joint(self) -> np.ndarray:
        """Normalised joint posterior, shape (33, 25, 15)."""
        return np.exp(self._log_post)

    def copy(self) -> "Posterior":
        other = Posterior.__new__(Posterior)
        other.theta_e, other.theta_i, other.beta = self.theta_e, self.theta_i, self.beta
        other._te, other._ti, other._b = self._te, self._ti, self._b
        other._log_post = self._log_post.copy()
        return other

    def _theta_axis(self, trait: TraitName) -> np.ndarray:
        return self._te if trait == "theta_e" else self._ti

    def log_lik(self, price: float, trait: TraitName, engaged: bool) -> np.ndarray:
        """Log-likelihood of one response across the whole joint grid."""
        return log_choice_lik(price, engaged, self._theta_axis(trait), self._b)

    def update(
        self,
        price: float,
        trait: TraitName,
        engaged: bool,
        weight: float = 1.0,
    ) -> None:
        """Multiplicative update, then renormalise. SPEC §3.4.

        ``weight`` supports the argument screen (SPEC §10), which refits with a
        disputed decision down-weighted to 0.5 (``situation``) or 0.2
        (``misread``). weight=1.0 is the ordinary in-play update.
        """
        self._log_post = self._log_post + weight * self.log_lik(price, trait, engaged)
        self._log_post -= _logsumexp(self._log_post)

    # -- summaries ----------------------------------------------------------

    def marginal(self, trait: TraitName) -> np.ndarray:
        joint = self.joint
        return joint.sum(axis=(1, 2)) if trait == "theta_e" else joint.sum(axis=(0, 2))

    def beta_marginal(self) -> np.ndarray:
        return self.joint.sum(axis=(0, 1))

    def mean_sd(self, trait: TraitName) -> tuple[float, float]:
        grid = self.theta_e if trait == "theta_e" else self.theta_i
        return _mean_sd(grid, self.marginal(trait))

    def beta_mean_sd(self) -> tuple[float, float]:
        return _mean_sd(self.beta, self.beta_marginal())

    def trait_beta_marginal(self, trait: TraitName) -> np.ndarray:
        """The (θ, β) marginal the ADO EIG runs on. Exact, not an
        approximation: the choice likelihood is constant along the other
        trait's axis. `rt_posterior.BehaviouralPosterior` exposes the same
        method, so `ado.py` works with either."""
        joint = self.joint
        return joint.sum(axis=1) if trait == "theta_e" else joint.sum(axis=0)

    def entropy(self) -> float:
        """Shannon entropy of the joint, in nats. Used by the ADO EIG (SPEC §5.3)."""
        p = self.joint
        nz = p > 0.0
        return float(-np.sum(p[nz] * np.log(p[nz])))

    def predicted_engage(self, price: float, trait: TraitName) -> float:
        """Posterior predictive P(engage), marginalising over (θ, β).

        SCHEMA §1: this must be recorded *before* the choice is known. That is
        what makes the prediction ledger honest.
        """
        z = self._b * (price - self._theta_axis(trait))
        return float(np.sum(self.joint * np.exp(-np.logaddexp(0.0, z))))

    def snapshot(self) -> PosteriorSnapshot:
        e_mean, e_sd = self.mean_sd("theta_e")
        i_mean, i_sd = self.mean_sd("theta_i")
        b_mean, b_sd = self.beta_mean_sd()
        return PosteriorSnapshot(
            theta_e_mean=e_mean,
            theta_e_sd=e_sd,
            theta_e_grid=self.theta_e.tolist(),
            theta_e_marginal=self.marginal("theta_e").tolist(),
            theta_i_mean=i_mean,
            theta_i_sd=i_sd,
            theta_i_grid=self.theta_i.tolist(),
            theta_i_marginal=self.marginal("theta_i").tolist(),
            beta_mean=b_mean,
            beta_sd=b_sd,
        )

    # -- refit --------------------------------------------------------------

    @classmethod
    def from_observations(
        cls,
        observations: Iterable[tuple[float, TraitName, bool]],
        weights: Sequence[float] | None = None,
    ) -> "Posterior":
        """Rebuild from scratch. Backs the SPEC §10 refit, which re-runs the
        whole sequence with one decision down-weighted rather than trying to
        divide it back out."""
        obs = list(observations)
        w = [1.0] * len(obs) if weights is None else list(weights)
        if len(w) != len(obs):
            raise ValueError(f"weights length {len(w)} != observations length {len(obs)}")
        post = cls()
        for (price, trait, engaged), wi in zip(obs, w):
            post.update(price, trait, engaged, weight=wi)
        return post


def _logsumexp(a: np.ndarray) -> float:
    m = float(np.max(a))
    if not np.isfinite(m):
        return m
    return m + float(np.log(np.sum(np.exp(a - m))))


def _mean_sd(grid: np.ndarray, marginal: np.ndarray) -> tuple[float, float]:
    mean = float(np.sum(grid * marginal))
    var = float(np.sum(marginal * (grid - mean) ** 2))
    return mean, float(np.sqrt(max(var, 0.0)))

File: PriorsEngine/Sources/PriorsEngine/Posterior.swift
//  Posterior.swift
//  Grid posterior over (θ_e, θ_i, β). SPEC §3.4.
//
//      P(engage | p, θ, β) = 1 / (1 + exp(β · (p − θ)))
//
//  Joint over 33 × 25 × 15 = 12,375 cells. Multiplicative update after every
//  decision, renormalise. That is the whole model.
//
//  Port of `priors-research/priors/posterior.py`, which is validated against
//  ADOpy. `PosteriorGoldenTests` asserts this file reproduces that reference on
//  a fixed decision sequence, so the two must not drift.
//
//  Everything is done in log space: β reaches 30 and prices reach 0.85, so
//  β·(p−θ) reaches ~24 and a naive exp() loses precision well before that.

import Foundation

/// Numerically stable `log(sigmoid(-z))`, i.e. `-log(1 + exp(z))`.
@inlinable
func negLogAddExpZero(_ z: Double) -> Double {
    z >= 0 ? -(z + log1p(exp(-z))) : -log1p(exp(z))
}

public struct Posterior: Sendable, ChoicePosterior {
    public let thetaE: [Double]
    public let thetaI: [Double]
    public let beta: [Double]

    /// Flat log-posterior. Index order is `(e, i, b)` — the same axis order as
    /// the Python reference, and load-bearing for the golden fixtures.
    public private(set) var logPost: [Double]

    @usableFromInline let nE: Int
    @usableFromInline let nI: Int
    @usableFromInline let nB: Int

    public init() {
        thetaE = Grids.thetaE.values()
        thetaI = Grids.thetaI.values()
        beta = Grids.beta.values()
        nE = thetaE.count
        nI = thetaI.count
        nB = beta.count
        // SPEC §3.1/§3.2 — uniform prior over each grid's support.
        let count = nE * nI * nB
        logPost = [Double](repeating: -log(Double(count)), count: count)
    }

    @inlinable
    public func index(_ e: Int, _ i: Int, _ b: Int) -> Int { (e * nI + i) * nB + b }

    // MARK: - Update

    /// Multiplicative update, then renormalise. SPEC §3.4.
    ///
    /// `weight` supports the argument screen (SPEC §10), which refits with a
    /// disputed decision down-weighted to 0.5 (`situation`) or 0.2 (`misread`).
    /// `weight` 1.0 is the ordinary in-play update.
    public mutating func update(price: Double, trait: Trait, engaged: Bool, weight: Double = 1.0) {
        // The likelihood is constant along the other trait's axis, so it is
        // computed once per (θ, β) pair rather than per cell.
        let theta = trait == .thetaE ? thetaE : thetaI
        var table = [Double](repeating: 0, count: theta.count * nB)
        for t in 0..<theta.count {
            for b in 0..<nB {
                let z = beta[b] * (price - theta[t])
                table[t * nB + b] = weight * (engaged ? negLogAddExpZero(z) : negLogAddExpZero(-z))
            }
        }

        if trait == .thetaE {
            for e in 0..<nE {
                for i in 0..<nI {
                    let base = (e * nI + i) * nB
                    for b in 0..<nB { logPost[base + b] += table[e * nB + b] }
                }
            }
        } else {
            for e in 0..<nE {
                for i in 0..<nI {
                    let base = (e * nI + i) * nB
                    for b in 0..<nB { logPost[base + b] += table[i * nB + b] }
                }
            }
        }
        normalise()
    }

    private mutating func normalise() {
        let m = logPost.max() ?? 0
        guard m.isFinite else { return }
        var sum = 0.0
        for v in logPost { sum += exp(v - m) }
        let offset = m + log(sum)
        for k in 0..<logPost.count { logPost[k] -= offset }
    }

    /// Rebuild from scratch. Backs the SPEC §10 refit, which re-runs the whole
    /// sequence with one decision down-weighted rather than dividing it back out.
    public static func from(
        observations: [(price: Double, trait: Trait, engaged: Bool)],
        weights: [Double]? = nil
    ) -> Posterior {
        let w = weights ?? [Double](repeating: 1.0, count: observations.count)
        precondition(w.count == observations.count,
                     "weights count \(w.count) != observations count \(observations.count)")
        var p = Posterior()
        for (o, wi) in zip(observations, w) {
            p.update(price: o.price, trait: o.trait, engaged: o.engaged, weight: wi)
        }
        return p
    }

    // MARK: - Summaries

    public var joint: [Double] { logPost.map(exp) }

    public func marginal(_ trait: Trait) -> [Double] {
        let n = trait == .thetaE ? nE : nI
        var out = [Double](repeating: 0, count: n)
        for e in 0..<nE {
            for i in 0..<nI {
                let base = (e * nI + i) * nB
                var s = 0.0
                for b in 0..<nB { s += exp(logPost[base + b]) }
                out[trait == .thetaE ? e : i] += s
            }
        }
        return out
    }

    public func betaMarginal() -> [Double] {
        var out = [Double](repeating: 0, count: nB)
        for e in 0..<nE {
            for i in 0..<nI {
                let base = (e * nI + i) * nB
                for b in 0..<nB { out[b] += exp(logPost[base + b]) }
            }
        }
        return out
    }

    /// The (θ, β) marginal the ADO EIG runs on. Exact, not an approximation:
    /// the choice likelihood is constant along the other trait's axis.
    public func traitBetaMarginal(_ trait: Trait) -> [Double] {
        let n = trait == .thetaE ? nE : nI
        var out = [Double](repeating: 0, count: n * nB)
        for e in 0..<nE {
            for i in 0..<nI {
                let base = (e * nI + i) * nB
                let row = (trait == .thetaE ? e : i) * nB
                for b in 0..<nB { out[row + b] += exp(logPost[base + b]) }
            }
        }
        return out
    }

    public func meanSD(_ trait: Trait) -> (mean: Double, sd: Double) {
        meanAndSD(grid: trait == .thetaE ? thetaE : thetaI, marginal: marginal(trait))
    }

    public func betaMeanSD() -> (mean: Double, sd: Double) {
        meanAndSD(grid: beta, marginal: betaMarginal())
    }

    /// Shannon entropy of the joint, in nats. Used by the ADO EIG (SPEC §5.3).
    public func entropy() -> Double {
        var h = 0.0
        for v in logPost {
            let p = exp(v)
            if p > 0 { h -= p * v }
        }
        return h
    }

    /// Posterior predictive P(engage), marginalising over (θ, β).
    ///
    /// SCHEMA §1: this must be recorded *before* the choice is known. That is
    /// what makes the prediction ledger honest.
    public func predictedEngage(price: Double, trait: Trait) -> Double {
        let theta = trait == .thetaE ? thetaE : thetaI
        let w = traitBetaMarginal(trait)
        var total = 0.0
        for t in 0..<theta.count {
            for b in 0..<nB {
                let z = beta[b] * (price - theta[t])
                total += w[t * nB + b] * exp(negLogAddExpZero(z))
            }
        }
        return total
    }

    public func snapshot() -> PosteriorSnapshot {
        let (me, se) = meanSD(.thetaE)
        let (mi, si) = meanSD(.thetaI)
        let (bm, bs) = betaMeanSD()
        return PosteriorSnapshot(
            thetaEMean: me, thetaESD: se,
            thetaEGrid: thetaE, thetaEMarginal: marginal(.thetaE),
            thetaIMean: mi, thetaISD: si,
            thetaIGrid: thetaI, thetaIMarginal: marginal(.thetaI),
            betaMean: bm, betaSD: bs
        )
    }
}

@usableFromInline
func meanAndSD(grid: [Double], marginal: [Double]) -> (mean: Double, sd: Double) {
    var mean = 0.0
    for k in 0..<grid.count { mean += grid[k] * marginal[k] }
    var variance = 0.0
    for k in 0..<grid.count {
        let d = grid[k] - mean
        variance += marginal[k] * d * d
    }
    return (mean, (max(variance, 0)).squareRoot())
}
